Autohide: optionally hide when the mouse is not over panel or indicator menus

quequotion opened the pull request:
I think it is worth [reconsidering](https://bugs.launchpad.net/wingpanel/+bug/1079575) autohide for [wingpanel](https://www.bountysource.com/issues/955498-add-optional-hide-mode) as an optional feature.

In fact many applications do not "fullscreen" when told to "maximize" and the presence of the panel enforces a 24px distinction between the two. There are also applications that attempt to draw actionable areas under the panel. Users could enable autohide to have their full resolution available and work around apps that do not play well with panels.

This patch adds an autohide mode and gsetting (org.pantheon.desktop.wingpanel autohide) with the behavior OFF as default.

quequotion commented:
We should probably talk about the animation speed. The initial draw for default behavior is set to 300ms, which is fine since it only happens once, but to work effectvely an autohiding panel should move faster. For autohide, I've set hide and unhide animations at 100ms. I'm not sure if adding gsetting for this would be convenience or clutter.

quequotion commented:
>over aggressive
It isn't intellihide, it is a dumb autohide. It's not so sexy, but the same behavior is offered for plank and I use it there as well. I think dumb autohide is what people who are going to the trouble of enabling an off-by-default authohide mode are more likely to want. Would it feel less aggressive if I added a grace period before hiding? Do you want a gsetting to specify the animation time?
>buggy
What do you mean by this? I just updated to fix some quirks; you might want to take another look.

cassidyjames commented:
The animation speed is not the problem (though 300ms is a pretty standard value in the OS; take a look at Plank, for example). The problems as I see them:

1. This is currently buggy. It hides seemingly randomly when my mouse is moving across the panel, or sometimes gets into a glitchy hide-unhide loop when my mouse is in certain places.

2. I don't see the value in straight up autohide; if nothing is open or in the way, hiding the panel has no functional value and only serves to hide UI elements that have been designed to be always-present.

3. Overlapping windows breaks the rules of things not being underneath the panel; instead, a hiding panel would need to adjust stuts and move maximized windows down when unhiding.

4. Working around and explicitly catering to poor-behaving apps is not something we do in elementary OS; apps that are well designed (especially ones designed for elementary OS) should behave in a way that makes sense with the platform.

5. I'm not a fan of adding hidden features. If it's a feature valuable enough to spend time developing, testing, and maintaining in perpetuity, we should be comfortable letting users select it. If we're not comfortable letting users select it, we shouldn't carry it.

quequotion commented:
>It hides seemingly randomly when my mouse is moving across the panel

Seems to be when transitioning from one indicator area another (ie, when moving from slingshot to the clock; when moving from the clock to the system indicators and vice versa). Odd; the evbox should cover the whole panel, but it seems to be split into three. Reminds me that one of the things I used to love about super-wingpanel was that it could autohide each area independently. I'll look into it.

>if nothing is open or in the way, hiding the panel has no functional value

Functional, no. It has aesthetic value. I really don't need anything being visible that isn't actively being used, and the panel will always be a mouse flick away anyway. I know that's just my personal preference, which is another reason the behavior is default OFF. Am I alone? Do I have OCD?

>a hiding panel would need to adjust stuts

This can be done. I'm not against shipping a version with that, but I would add an aditional gsetting for "floating" mode. I can see how the unhide trigger area (top 1px of the screen) could get in the way of window decorations with sloppy mouse operation; for my preference I'd rather the decorations always stay in the same place and the windows always stay the same size and shape.

>apps that are well designed (especially ones designed for elementary OS) should behave in a way that makes sense with the platform

Because many applications will never reach that standard, the solution here is a hide mode that recognizes when a window is "maximized" and then hides the panel. Ideally I'd like to do that in a window-manager agnostic way (ie, not rely on Gala specifically to inform wingpanel of window states) but I doubt such a proposal would be accepted (because supporting alternate window managers is not in your purview). Dumb autohide is a compromise that allows having autohide with wingpanel in alternate window managers.

>I'm not a fan of adding hidden features.

It might be worthwhile to add a tab for "Wingpanel" (or "Top Panel") to Switchboard's "Desktop" or "Tweaks" plugins, _if_ additonal features like autohide/hide on maximize will be approved.

quequotion commented:
The next update splits autohide into two modes: "Autohide" which will draw struts and push windows out of its space, and "Floating", which will overlay the top 24px.

I've changed the gsetting (org.pantheon.desktop.wingpanel autohide) to a string with three choices: 'Autohide', 'Floating', and 'Disabled', with 'Disabled' as default.

Redrawing the struts looks _really_ clunky, but it works.

If I can find a window-manager agnostic way to implement "hide on maximize", I'll add it.

cassidyjames commented:
>  It has aesthetic value. I really don't need anything being visible that isn't actively being used

This is why we do a lot of smart automatic coloring to make the panel blend in, whether it's to the display bezel (when apps are maximized) or the wallpaper (where it goes transparent and the foreground changes to contrast). It's both aesthetically pleasing _and_ always visible. We use the panel as an anchor point so users always have a way to get to the most important system functions.

>I've changed the gsetting (org.pantheon.desktop.wingpanel autohide) to a string with three choices: 'Autohide', 'Floating', and 'Disabled', with 'Disabled' as default.

And now this is passing even more design decisions off to the user. :stuck_out_tongue: If any sort of hiding is to be accepted, I think we need to figure out how exactly we want it to work, then implement and test it. Adding yet-another-option is not typically the correct course.

>It might be worthwhile to add a tab for "Wingpanel" (or "Top Panel") to Switchboard's "Desktop" or "Tweaks" plugins, if additonal features like autohide/hide on maximize will be approved.

Tweaks is not a first-party thing, so that's still definitely a hidden feature. I've started a branch for a Panel tab under the Desktop settings for a different feature, so that would be a natural place to put it. If we're comfortable committing to accepting and maintaining the feature (and added complexity) in perpetuity.

>Redrawing the struts looks really clunky, but it works.

I took a look at this. I agree it's not perfect, but I do think it's more ideal than overlapping content. GTK windows typically resize more smoothly than I'm seeing here, so I wonder if it could be better optimized.

>If I can find a window-manager agnostic way to implement "hide on maximize", I'll add it.

Check what Plank does here. Though it is likely to still be X-specific.

cassidyjames commented:
It looks like clicking outside of an indicator to close it is broken by this branch.

cassidyjames commented:
It also looks like this doesn't listen to the GSetting live. Settings should apply instantly and not require a restart of the process.

quequotion commented:
>It's both aesthetically pleasing and always visible.

I use wingpanel in particular because it can have a transparent background and blend in to any desktop. In fact, I [maintain a fork](https://github.com/quequotion/pantheon-bzr-qq/tree/master/wingpanel-standalone-git) that drops all those bells and whistles in favor of window-manager agnosticism and transparency. Nonetheless, I'd like to have dumb autohide available, just ask Plank has several autohide modes available, for user's preferences.

>We use the panel as an anchor point so users always have a way to get to the most important system functions

My initial efforts at autohide in wingpanel would allow it initially appear and stay visible until the first leave_notify_event for this reason, to give users their bearings. A few hours ago I decided this was too quirky to present as a serious proposal for merging upstream.

>And now this is passing even more design decisions off to the user.

I get that you don't want a lot of settings for people to fiddle with. If drawing the struts weren't so clunky I might be comfortable giving up "floating" mode (and maintaining it in my fork), but for the time being my proposal includes both more for the sake of debate than intent to implement both.

>X-specific

One massive undertaking at a time.

>clicking outside of an indicator to close it is broken

I've been living under the assumption that this never worked in wingpanel for a very long time. It probably has to do with the evbox I'm using to catch {enter,leave}_notify_event; which might not have sufficent masks. Any advice on this is welcome.

>this doesn't listen to the GSetting live

I would very much like to implement this, but I do not know how. Again, any advice welcome.

quequotion commented:
The next step is to implement a hide on maximized (active window). This seems like it will be a big step.

- It's going to require another library, libwnck.
- This  will add a lot of complexity to src/PanelWindow.vala in particular

I think I can whittle down what was done with super-wingpanel, but there's no getting out of libwnck. It's already a dependency of plank? I think a pull request that  includes a new dependency is a bit of a long shot.

I see the trend of having a "HideManager" to do this for other components. I had in mind to implement the autohide directly in PanelWindow, because it was easy. Should I port to the common approach?

I would also like to take this opportunity to further extol upon the virtues of dumb autohide:

- requires no additional dependencies
- display server protocol agnostic
- [125 lines](https://github.com/elementary/wingpanel/pull/93/files).

donadigo commented:
@quequotion: you don't need libwnck. Also we try to stay from X dependent libraries right now to not introduce code that will not work when we move to Wayland.

Wingpanel already knows about the current window state, just think about it, how does it go black when a window is maximized?

If you need more control over the events that happen inside the compositor, you should look at the "wingpanel-interface" which is a gala plugin specifically made to exchange useful data between the wm and the panel. It is also contained within this repository.

As for hiding logic: ideally, all the show and hide logic should be in another class, throwing everything into the PanelWindow class means that the code will become less maintainable in the future.

surajmandalcell commented:
Well you guys might not appreciate the auto-hide feature as a design feature but as a user who is using 1366x768 window resolution (which most of the older laptops do run on) it really helps when you're trying to use all the space you can get. 
Also remember that plank doesn't appear when hovering the bottom edges while full-screen (win+f) which makes multitasking hard while full-screen.

donadigo commented:
@surajmandalcell we do and we are not opposed to implement that, but it's kind of hard to achieve a good working implementation with the panel, there are some blockers which make the feature laggy or random to use. I haven't looked at this branch in detail yet but guessing by the commit history it would be worth to take another look at how it works in a real world.

I've already attempted to implement such thing and it's not trivial to do this properly.

quequotion commented:
@surajmandalcell 

Thank you for your support! Just to let you know, I'm not giving up on this. I tend to work in spurts and it seems like I'm on my own, so it may take a while, but I will try to get the "dodge" type of autohide implemented eventually. I also need to move the autohiding into its own class, but to do so I need to learn more Vala.

@donadigo 

I would appreciate if you would look at it again; It has been while since I fixed several of the issues you brought up. I don't think this is actually such a difficult thing to implement, except that my programming skills are very rusty.

surajmandalcell commented:
@quequotion I can actually work on this if you want but the thing is i don't know vala more than changing little frontend parts so i need to learn vala too. My first semester just ended so i have a month of holiday in which i can work on these stuff. Do tell if i can help you out on some repetitive tasks or something similar.

And

Do you use the default valadoc.org to learn vala through something else?

quequotion commented:
@surajmandalcell 

Thanks for the offer! I don't have a more detailed tasklist than the discussion here; the next steps will be difficult ones. Yes, I use valadoc.org and also I spent a lot of time looking over the code of various Pantheon components. The elementary OS devs build their applications in a fairly consistent style, so it's easy to follow what goes where. Another useful resource is the [GTK+ 3 Reference Manual](https://developer.gnome.org/gtk3/stable/).

quequotion commented:
I've added an autohide delay, which defaults to 200ms; the same delay is used for both hide and unhide. I chose this number as a compromise between the original 300ms animation speed and the 100ms I consider practical for autohide. It is configurable via a gconf setting at io.elementary.desktop.wingpanel.delay (0~1000ms).

I've also modified the strut drawing behavior, so that they are drawn after the panel animates for "Autohide", "Disabled" and yet-to-be-implemented "Dodge". This changes the animation from a clunky push down / pull up to a quick knock down / jump up; I almost like it--but I still prefer "Float" as it doesn't move windows around.

quequotion commented:
Imported dodge-maximize code from Keith Bailey's [branch](https://code.launchpad.net/~kbailey4444/wingpanel/dodge-maximize/+merge/262278).

There are now _four_ autohide modes available; I do not intend to get all of them into wingpanel. I'd like some reviewers to [try them out and give me their opinion](https://plus.google.com/108178590046843354025/posts/ccEbydHhVVb) about which behaviors to keep. Dodge-Float is my favorite.

donadigo commented:
Here we go again:
Nice, getting close to something that could be merged! I personally like the Dodge-Float and still Float modes. The issues with Dodge and Autohide remain the same, when using these modes, the maximized window will reallocate it's size and the experience will not be a great one.

Issues:
* Don't use libwnck. It uses X server to gain the info about the windows while there's a much more native way - through the window manager. Wingpanel itself has a gala plugin that can expose that information to the wingpanel process.

* I feel like the wingpanel animation is still really slow (not talking about delay timeout). I often wait with my mouse for the panel to show.

* There's an issue with how struts currently work with notifications. If you try to use one of the float modes and try to increase / decrease volume with the scroll wheel, the volume notification appears overlaying the panel, stealing the mouse focus and in result, the panel hides. This will not be an easy one to fix.

More to come.

quequotion commented:
> * Don't use libwnck. It uses X server to gain the info about the windows while there's a much more native way - through the window manager.

I need wingpanel and this feature to work with other window managers, so libwnck seems to be the only option. If there were a display-server agnostic method other than using Gala I would go for that, but I am not aware of any. I understand that you put your distribution first, but the very reason I fell in love with (super-)wingpanel and plank was that they were modular and worked well with most window managers. Maybe later I'll make a gala-specific version of this, if that's what's required to get through this pull request, but to be honest I wouldn't feel good about contributing to walling in this garden.

> * I feel like the wingpanel animation is still really slow (not talking about delay timeout). I often wait with my mouse for the panel to show.

The animation I'm doing here comes from the original drawing animation (when wingpanel appears at startup). If this is slow, it is because the original model is slow. I'd appreciate any advice on speeding this up.

> * There's an issue with how struts currently work with notifications. If you try to use one of the float modes and try to increase / decrease volume with the scroll wheel, the volume notification appears overlaying the panel, stealing the mouse focus and in result, the panel hides. This will not be an easy one to fix.

The float modes do not draw struts. I use wingpanel with a different window manager and notifications daemon; the bubbles do appear over the panel but the panel retains focus, so the mousewheel can still change the volume. Why does gala allow notifications to steal focus?

By the way, apologies, I force pushed a hard reset to delete the commits I tried to get Travis CI to find libwnck. I don't know why it is having trouble; there is no problem on my end. I really hate cmake, I really really do.

quequotion commented:
The latest changes are probably the last until migrating this to its own service file, unless someone finds bugs to fix.

Responsiveness is greatly improved, as is behavior. The default 200ms delay is now quite practical: the panel will not show at all for brief, accidental contacts with the top edge of the screen and it's quite difficult to bounce the panel up and down by waggling the mouse along the top edge of the screen (it is possible with a delay less than 150ms, problematic with a delay less than 100ms). I'm tempted to hard code it at this point, but there could still be an argument for keeping the delay customizable for accessibility and it makes sense to leave it open as long as this pull request is in more of a technological demonstration form than whatever might eventually be accepted.

One significant change is that I removed animate_panel (), which was just a duplicate of update_panel_dimensions () minus the call to update_struts (). This was written at a time when my strategy was to avoid making changes to upstream code as much as possible, which led to many sub-optimal design choices. I'm thinking of a new plan for migrating the hiding code to its own service: It doesn't seem like this can be done without making some modifications to src/PanelWindow.vala anyway, so I will migrate some of its code along with the hiding logic, into something like Services/AnimationManager.vala.

That said, I still have a lot of vala to learn before any such thing is possible.

TonioGela commented:
Any news? It will be a nice feature to have!

quequotion commented:
@TonioGela I tend to work in spurts. Sometimes a while may pass between updates, but the work continues! Right now the to-do list is as follows:

1. Move animation routines out of PanelWindow.vala and into their own class
2. Reduce the number of available hide modes (the elementary OS devs only want "Dodge"; [the users mostly want "Autohide"](https://plus.google.com/108178590046843354025/posts/ccEbydHhVVb); I like "Dodge-Float" and "Float")
3. Rewrite Dodge to use Gala API instead of libwnck (although I'd rather keep it DE agnostic, the elementary OS devs have in mind that this will make porting to Wayland easier)

In the mean time, I could really use some help with Debian packaging. I'd like to ship this version of wingpanel in my [PPA](https://launchpad.net/~quequotion/+archive/ubuntu/qqdesk), but my last attempt failed to build for trying to compile gschemas.

n-osennij commented:
+1

n-osennij commented:
The timeframe?

quequotion commented:
@n-osennij I wish I had a timeframe for you; I know there's a lot of demand for this feature.
[This fork](https://github.com/quequotion/wingpanel) is in fact already working, but has not been released for elementary OS.
[The remaining to-do items](https://github.com/elementary/wingpanel/pull/93#issuecomment-434031813) are to qualify for merging.
It would release some of the pressure if someone could help me release a PPA package of this in the interim. I seriously have more difficulty making Debian packages than writing programs.

In the mean time, if you are comfortable using git, compiling, and installing software without a package manager (`apt` in elementary OS/Ubuntu/Debian), please try this out if you haven't already!

InvisibleRasta commented:
i tried building this but i get an error when running meson build --prefix=/usr
Meson encountered an error in file lib/meson.build, line 20, column 0:
Custom_install_dir must be a string

quequotion commented:
> i tried building this but i get an error when running meson build --prefix=/usr
> Meson encountered an error in file lib/meson.build, line 20, column 0:
> Custom_install_dir must be a string

That file is [not modified in this fork](https://github.com/elementary/wingpanel/pull/93/files) and `meson build --prefix=/usr` works for me. If meson isn't working for you, the package can still be built with cmake:
```
mkdir build && cd build
cmake .. \
    -DCMAKE_BUILD_TYPE='Release' \
    -DCMAKE_INSTALL_PREFIX='/usr' \
    -DCMAKE_INSTALL_LIBDIR='lib' \
    -DGSETTINGS_COMPILE='FALSE'
make
```
Thanks for trying it out! Let me know if you have any more trouble.

quequotion commented:
I spent a few months getting used to "Dodge-Float" (autohide on fullscreen/maximized).

The mechanism works: when the focused window is fullscreen or maxmized, wingpanel autohides.
When the focused window is small, or no window is focused, wingpanel appears--and with "Float", the panel also appears on mouseover (without strutting).

- Sometimes a fullscreen or maximized window will make a popup window (modal dialog, etc); and wingpanel appears (focus change to subordinate window); if you don't recognize the dialog, and close it, it seems like the panel should not be open (a maximized window is under it, and also under that dialog).* 
- "Float" is a complete autohide mechanism itself; the "Dodge" modes are a kind of "Autoshow".**

*Could be worked around by changing "hide when focused window on this desktop is fullscreen/maximized" to "hide when _any_ window on this desktop is fullscreen/maximized", at the cost keeping the panel hidden when _conciously_ using a smaller window on top of a maximized or fullscreen one.
**Still feels more natural

lhorace commented:
Hello there,

I'm testing wingpanel-standalone-git on ArchLinux here. One issue, someone mentioned that when a Window is maxmize, Wingpanel should go dark? Well using the Dodge-float feature, the readability of the text is difficult? Am I missing something here?

Regards

quequotion commented:
@lhorace 
Yes, the [AUR pacakge](https://aur.archlinux.org/cgit/aur.git/tree/?h=wingpanel-standalone-git) has additional patches to remove features that depend on the Gala window manager, such as the adaptive background color, leaving the panel background permanently transparent.

If you are using Gala, you could rebuild the package with those features by commenting out in the patching in the PKGBUILD like so:

```
  #Standalone patches
  #msg2 "Remove Gala dependecies"
  #rm -rf wingpanel-interface
  #rm src/Services/BackgroundManager.vala
  #patch -Np2 < ../minus-backgroundmanager.patch
  #patch -Np2 < ../minus-galaplugin.patch
  #patch -Np2 < ../minus-gala.patch

  #Cogl can't be found when not using gala's cmake package; wtf?
  #msg2 "Remove CoglFixes (broken)"
  #patch -Np2 < ../y-is-broken-cogl.patch
```

n-osennij commented:
Any news?

quequotion commented:
> Any news?

Nice to know there is still interest!

I'll get back to work sooner or later. For now this is working as much as I need it to be. Not distributed well though.

I think it would relieve some of the pressure if it were easy to install as a PPA package or flatpack, but I am not competent in making those kinds of packages. Could use some help there.

n-osennij commented:
> > Any news?
> 
> Nice to know there is still interest! 

The absence of automatic hiding is the function why I do not use this system as the main one.

> I'll get back to work sooner or later. For now this is working as much as I need it to be. Not distributed well though.
> 
> I think it would relieve some of the pressure if it were easy to install as a PPA package or flatpack, but I am not competent in making those kinds of packages. Could use some help there.

I am not competent in this. I can not help.

aswinmohanme commented:
This is a well-needed feature particularly on smaller screens, I have been using the wingpanelautohide package and it has improved my experience in elementary a lot.

quequotion commented:
@aswinmohanme 

> This is a well-needed feature particularly on smaller screens
Yes, I strongly agree. Desktop real-estate is precious; 24px are important.

>I have been using the wingpanelautohide package
To which package are you referring?

aswinmohanme commented:
> To which package are you referring?

The `elementary-wingautohide` package from `ppa:yunnxx/elementary`

donadigo commented:
@quequotion I have submitted my own version here: https://github.com/elementary/wingpanel/pull/212. If you want you could incorporate some of my changes into your own fork as well.

n-osennij commented:
any updates?

CristianGustav commented:
Plesae explain me someone how to install this panel

quequotion commented:
Thank you for your interest! 

Explain [it to] you:
_short version_:
```
git clone https://github.com/quequotion/wingpanel.git
cd wingpanel
meson build
ninja build
ninja -C build install
```

_long version_:
This is not available as a debian package; it cannot be installed by apt. It is available as [an AUR package](https://aur.archlinux.org/packages/wingpanel-standalone-git/) for Archlinux (and derivatives). If you do not use Archlinux or a derivative operating system, (remove your existing wingpanel package and) build and install from git. See instructions above.

Note, [donadigo's version](ttps://github.com/elementary/wingpanel/pull/212) is more likely to become part of Elementary OS in the future (using their libgala to interact with the desktop and windows, while mine uses [gnome's libwnck](https://developer.gnome.org/libwnck/stable/overview.html)). Installing it is more complicated at the moment (depends on [an also wip version of gala](https://github.com/elementary/gala/pull/535); more git commands (it is a branch of the official wingpanel repository)).

laxit-patel commented:
this blog provides niche overall details about hiding wingpanel in JUNO
http://bepoweruser.com/2019/09/09/hide-wingpanel-in-elementary-os-juno/

quequotion commented:
@laxit-patel 

>so i just did this with simple killall and start

This is not _hiding_ wingpanel, it's killing it and restarting it. That's probably not going to get along with `cerbere` for one, and is a generally over-the-top, heavy, unsound way of doing things. Hence this pull request, and donadigo's, adding autohide to wingpanel itself. 

Not to disparage its author, it is one kind of way around the problem, but better can be done.

quequotion commented:
Rebased for cleanup! I've split this into two commits: one for Autohide and Float, which do not require any additional libraries, and a second for Dodge and Dodge-Float, for which I am using libwnck.

st3f4nm commented:
> Thank you for your interest!
> 
> Explain [it to] you:
> _short version_:
> 
> ```
> git clone https://github.com/quequotion/wingpanel.git
> cd wingpanel
> meson build
> ninja build
> ninja -C build install
> ```
> 
> _long version_:
> This is not available as a debian package; it cannot be installed by apt. It is available as [an AUR package](https://aur.archlinux.org/packages/wingpanel-standalone-git/) for Archlinux (and derivatives). If you do not use Archlinux or a derivative operating system, (remove your existing wingpanel package and) build and install from git. See instructions above.
> 
> Note, [donadigo's version](https://github.com/elementary/wingpanel/pull/212) is more likely to become part of Elementary OS in the future (using their libgala to interact with the desktop and windows, while mine uses [gnome's libwnck](https://developer.gnome.org/libwnck/stable/overview.html)). Installing it is more complicated at the moment (depends on [an also wip version of gala](https://github.com/elementary/gala/pull/535); more git commands (it is a branch of the official wingpanel repository; etc)).

Hi quequotion,  I'm trying to install this on elementary OS 5.1 but im getting errors when doing the "meson build"

Dependency libmutter found: NO (tried pkgconfig and cmake)
Dependency libmutter found: NO (tried pkgconfig and cmake)
Dependency libmutter found: NO (tried pkgconfig and cmake)
Dependency libmutter-0 found: NO (tried pkgconfig and cmake)
Dependency libmutter-1 found: NO (tried pkgconfig and cmake)
Dependency libmutter-2 found: YES (cached)
Dependency libmutter-2 found: YES (cached)
Dependency mutter-cogl-2 found: YES (cached)
Dependency mutter-cogl-pango-2 found: YES (cached)
Dependency mutter-cogl-path-2 found: YES (cached)
Dependency mutter-clutter-2 found: YES (cached)
Dependency libmutter-3 found: NO (tried pkgconfig and cmake)
Dependency libmutter-4 found: NO (tried pkgconfig and cmake)
Dependency libmutter-5 found: NO (tried pkgconfig and cmake)
Dependency libmutter-6 found: NO (tried pkgconfig and cmake)

I'm sorry, I'm new to this.  how do I fix those libmutter dependency issues?

donadigo commented:
@st3f4nm these "errors" aren't critical, it's just the build system printing what mutter package you have installed (in this case libmutter-2). I'd recommend following the next instructions anyway and seeing if they succeed.

st3f4nm commented:
> @st3f4nm these "errors" aren't critical, it's just the build system printing what mutter package you have installed (in this case libmutter-2). I'd recommend following the next instructions anyway and seeing if they succeed.

Thank you! It worked.  haha.  Sorry for being a newbie on this.  

Thanks @donadigo and @quequotion

n-osennij commented:
s it's die again?

quequotion commented:
> so it's die again?

This has never been dead.

Solar-Lancelo commented:
I think its helpful if the wing-panel can get autohide.
I have setted the hot corner to open application menu.It's much more convenient than clicking the "application" at left end of wing-panel. Thus, the only use of wing-panel it's just shutdown or logout.

Solar-Lancelo commented:
It's not useful to take place of the top screen especially in 14-inch laptop

AbhinavMadahar commented:
> Thank you for your interest!
> 
> Explain [it to] you:
> _short version_:
> 
> ```
> git clone https://github.com/quequotion/wingpanel.git
> cd wingpanel
> meson build
> ninja build
> ninja -C build install
> ```
> 
> _long version_:
> This is not available as a debian package; it cannot be installed by apt. It is available as [an AUR package](https://aur.archlinux.org/packages/wingpanel-standalone-git/) for Archlinux (and derivatives). If you do not use Archlinux or a derivative operating system, (remove your existing wingpanel package and) build and install from git. See instructions above.
> 
> Note, [donadigo's version](https://github.com/elementary/wingpanel/pull/212) is more likely to become part of Elementary OS in the future (using their libgala to interact with the desktop and windows, while mine uses [gnome's libwnck](https://developer.gnome.org/libwnck/stable/overview.html)). Installing it is more complicated at the moment (depends on [an also wip version of gala](https://github.com/elementary/gala/pull/535); more git commands (it is a branch of the official wingpanel repository; etc)).

This worked correctly for me, but I first had to install `libwnck-3-dev` with APT. Also, the `ninja build` command failed on me, giving me the error
```sh
ninja: error: loading 'build.ninja': No such file or directory
```
but I just ignored it and moved on to the next command. After that, I could use dconf to make the wingpanel hide on maximisation.

quequotion commented:
@AbhinavMadahar Thank you for your interest!

Yes, libwnck is a dependency. I haven't given specific instructions on how to install it as they would be different for each distribution (this can be used outside of elementary OS), but I should find a way to ammend that it's required to those instructions.

I'm not responsible for the build system, but it sounds like it isn't a real problem. I've just updated against the latest upstream changes, maybe it's fixed?

AbhinavMadahar commented:
Yeah, I don't think any change is needed. I just wanted to document this in
case someone else has a similar issue.

On Fri, 3 Jan, 2020, 9:04 PM Que Quotion, <notifications@github.com> wrote:

> @AbhinavMadahar <https://github.com/AbhinavMadahar> Thank you for your
> interest!
>
> Yes, libwnck is a dependency. I haven't given specific instructions on how
> to install it as they would be different for each distribution (this can be
> used outside of elementary OS), but I should find a way to ammend that it's
> required to those instructions.
>
> I'm not responsible for the build system, but it sounds like it isn't a
> real problem. I've just updated against the latest upstream changes, maybe
> it's fixed?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elementary/wingpanel/pull/93?email_source=notifications&email_token=AB7NMEHCCVJD5VYGQNQWTM3Q37VD7A5CNFSM4ETMUU72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEICOVZQ#issuecomment-570747622>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7NMEFIN5S7H5QVKONRRE3Q37VD7ANCNFSM4ETMUU7Q>
> .
>

n-osennij commented:
try to install and got next error

```
autumn@autumn-VirtualBox:~/wingpanel$ meson build
The Meson build system
Version: 0.49.2
Source dir: /home/autumn/wingpanel
Build dir: /home/autumn/wingpanel/build
Build type: native build
Project name: wingpanel
Project version: 2.2.6
Native C compiler: cc (gcc 7.4.0 "cc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
Native Vala compiler: valac (valac 0.40.19)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Dependency glib-2.0 found: YES 2.56.4
Dependency gio-2.0 found: YES 2.56.4
Dependency gio-unix-2.0 found: YES 2.56.4
Dependency gmodule-2.0 found: YES 2.56.4
Dependency gtk+-3.0 found: YES 3.22.30
Found CMake: NO
Dependency libwnck-3.0 found: NO (tried pkgconfig)

meson.build:34:0: ERROR:  Dependency "libwnck-3.0" not found, tried pkgconfig

A full log can be found at /home/autumn/wingpanel/build/meson-logs/meson-log.txt
autumn@autumn-VirtualBox:~/wingpanel$ sudo apt install libwnck-3.0
Чтение списков пакетов… Готово
Построение дерева зависимостей       
Чтение информации о состоянии… Готово
Заметьте, выбирается «libwnck-3-0» для регулярного выражения «libwnck-3.0»
Уже установлен пакет libwnck-3-0 самой новой версии (3.24.1-2).
Следующие пакеты устанавливались автоматически и больше не требуются:
  gir1.2-vte-2.91 libido3-0.1-0
Для их удаления используйте «sudo apt autoremove».
Обновлено 0 пакетов, установлено 0 новых пакетов, для удаления отмечено 0 пакетов, и 0 пакетов не обновлено.
autumn@autumn-VirtualBox:~/wingpanel$
```

quequotion commented:
> try to install and got next error. I install `sudo apt install libwnck-3-0` but in not solve problem

Thank Debian / Ubuntu for their crazy packaging concept. You need `libwnck-3-dev`.

miere commented:
Hi guys, any thoughts?

This is a must have feature for 11", 13" and 14" notebooks. Although the new ones have higher screen resolution, they still have small screen.

quequotion commented:
@miere This is fully working, just not conveniently packaged.

If anyone could help me figure out how to package it to distribute on launchpad it would be a big help.

xXAquaNightXx commented:
Where do i got to change the settings and enable the autohide after installing?

quequotion commented:
> Where do i [go] to change the settings and enable the autohide after installing?

They are gsettings (dconf): change them with the terminal client, `gsettings`, or the graphical client, `dconf-editor`. The path is provided in the top post of this PR.

AngelDMA commented:
@quequotion Hi, I just installed elementary and I was looking for this, any way, When I try meson build, this give me some errors 

Dependency libmutter-2 found: NO (tried pkgconfig and cmake)
Dependency libmutter-3 found: NO (tried pkgconfig and cmake)
Dependency libmutter-4 found: NO (tried pkgconfig and cmake)
Dependency libmutter-5 found: NO (tried pkgconfig and cmake)
Dependency libmutter-6 found: NO (tried pkgconfig and cmake)

wingpanel-interface/meson.build:58:1: ERROR:  Problem encountered: No supported mutter library found!

What should I do?

AngelDMA commented:
Oh sorry, I just checked the dependencies

n-osennij commented:
so why it is not still merged?

quequotion commented:
@n-osennij two elementary OS developers have filed requests for changes. I haven't heard from @cassidyjames about it lately, so I've filed a re-request for review. Although the performance and behavior issues are, to my knowledge, fixed, I don't expect the fate of this pull request to change.

1. My "Dodge" implementation depends on libwnck, which does not support Wayland; last I checked the elementary OS devs are planning on Pantheon becoming Wayland compatible, although plank also depends on libwnck at the moment. This is why @donadigo is working on another fork using libgala for "Dodge".

2. This pull request features too many autohide modes. This is in part a technology demonstration for the different approaches. In order to prep this for adoption, I'd have to rebase with just one or two modes (likely "Float" and "Dodge-Float").

3. I haven't gotten around to recoding this in a more "elementary OS" standard. Looking at how other apps are laid out, I should probably split the code for hiding (or all animation) into it's own file, ie `src/Services/AnimationManager.vala`. Prequisite to that would be learning a  lot more Vala.

What I really need is help with Debian packaging so I could distribute this fork to elementary OS users via PPA.

janxkoci commented:
**Question:**

You mentioned a couple of times that you want this to work for *fullscreen* windows too, not just maximized. So if some app has a slide-in toolbar in fulscreen mode activated by mousing to the top edge of the screen, how would this be handled? As I understand it, there would be quite a conflict between the app and the wingpanel.

Examples of apps with slide-in toolbar in fullscreen mode: 

- Evince (default pdf / djvu / eps reader in elementary OS),
- Epiphany (not the default elementary version, this was added in 3.30 or 3.32, which can be still installed using flatpak / snap),
- Firefox (this behaviour has been there for like ages),
- Totem (GNOME Videos),
- Eye of GNOME,

and probably a bunch more.

Any ideas how such conflicts should be handled? In my view the panel hiding-and-revealing should only apply to maximized windows, but not fullscreen windows (there just hide everything, as is currently the case).

quequotion commented:
@janxkoci There may not *be* a conflict. I haven't tested all of those applications, but I have tested firefox. It works out that Firefox's pulldown appears first. I added a customizable delay some time ago, although there was no conflict without it to begin with. The default delay is 200ms, and the unhide animation has a duration of 100ms.So long as other applications' pulldowns trigger faster (I suspect they have no delay) there shouldn't be a problem.

That said, if a user had accessibility issues or otherwise did not move the pointer for a while after a pulldown appeared, yes it is possible to have the pull down obscured (if using "Float" mode) when the panel unhides.This could be worked around with "Autohide" mode, which redraws struts and will push the full screen window out of the way (squish it downward 24px),but the animation isn't pretty.

mikeball commented:
I'm new to elementary and would like to try this pull request out. How would I go about this, and rollback if necessary?

quequotion commented:
@mikeball 

>How would I go about this
The instructions are linked at the top.

>and rollback if necessary?
In theory, you should be able to remove the installed files with `ninja uninstall` and then reinstall a standard wingpanel package.

simonfernandezprada commented:
> Thank you for your interest!
> 
> Explain [it to] you:
> 
> _short version_ *****:
> 
> ```
> git clone https://github.com/quequotion/wingpanel.git
> cd wingpanel
> meson build
> ninja build
> ninja -C build install
> ```
> 
> _long version_:
> 
> This is not available as a Debian package; it cannot be installed by `apt`. It is available as [an AUR package](https://aur.archlinux.org/packages/wingpanel-standalone-git/) for Archlinux (and derivatives). If you do not use Archlinux or a derivative operating system, (remove your existing wingpanel package and) build and install from git.
> 
> Note, [donadigo's version](https://github.com/elementary/wingpanel/pull/212) is more likely to become part of elementary OS in the future (using their libgala to interact with the desktop and windows, while mine uses [gnome's libwnck](https://developer.gnome.org/libwnck/stable/overview.html)). Installing donadigo's is more complicated at the moment (depends on [an also wip version of gala](https://github.com/elementary/gala/pull/535); more git commands (it is a branch of the official wingpanel repository; etc)).
> 
> ***** _libwnck3_ is an additional build dependency. In Debian (eg Ubuntu, eg elementary OS) [preinstall](https://github.com/elementary/wingpanel/pull/93/files#diff-90829e76e906f1c73140c7ded7e1b268) `libwnck-3-dev`.

I can't make it work. I seem to have installed everything but no dice.
meson was installed

`meson build
Directory already configured.

Just run your build command (e.g. ninja) and Meson will regenerate as necessary.
If ninja fails, run "ninja reconfigure" or "meson --reconfigure"
to force Meson to regenerate.

If build failures persist, run "meson setup --wipe" to rebuild from scratch
using the same options as passed when configuring the build.
To change option values, run "meson configure" instead.
chulz@chulz-Lenovo-B41-30:~/wingpanel$ ninja
ninja: error: loading 'build.ninja': No such file or directory
chulz@chulz-Lenovo-B41-30:~/wingpanel$ ninja build
ninja: error: loading 'build.ninja': No such file or directory
`

ninja was installed

`ninja -C build install
ninja: Entering directory `build'
[0/1] Installing files.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/mesonbuild/mesonmain.py", line 112, in run
    return options.run_func(options)
  File "/usr/lib/python3/dist-packages/mesonbuild/minstall.py", line 500, in run
    with open(os.path.join(log_dir, 'install-log.txt'), 'w') as lf:
PermissionError: [Errno 13] Permission denied: 'meson-private/../meson-logs/install-log.txt'
FAILED: meson-install 
/usr/bin/meson install --no-rebuild
ninja: build stopped: subcommand failed.
`

I don't know what else to do. If you read these lines, please share some light to this newbie. Thanks friend!

quequotion commented:
@simonfernandezprada 

You need more than meson. I have no idea about your error messages, but please see the [readme](https://github.com/quequotion/wingpanel#building-and-installation) for the full list of prerequisite packages.

It would help to know what operating system you are attempting to build this in (and what release of that OS).

_If you are using Debian, Ubuntu, or any derivative (such as Elementary OS)_, you probably need to start here:

` apt install -y libgala-dev libgee-0.8-dev libgirepository1.0-dev libglib2.0-dev libgranite-dev libgtk-3-dev meson valac libwnck-3-dev`

Then follow the instructions as given. I have not included this line explicitly because it is operating-system specific: these packages are not necessarily known by the same names by other operating systems' package managers.

DAC324 commented:
> It is available as [an AUR package](https://aur.archlinux.org/packages/wingpanel-standalone-git/) for Archlinux (and derivatives).
The Wingpanel-standalone-git package appears to be not maintained anymore, and it cannot be installed due to the following error:
<pre>
==> Starting prepare()...
  -> Autohide
patching file meson.build
Hunk #2 succeeded at 31 with fuzz 2.
patching file schemas/io.elementary.desktop.wingpanel.gschema.xml
patching file src/PanelWindow.vala
Hunk #4 succeeded at 305 (offset -1 lines).
patching file src/Widgets/Panel.vala
Hunk #2 succeeded at 94 with fuzz 1 (offset 9 lines).
patching file src/meson.build
Hunk #1 succeeded at 16 (offset 1 line).
  -> Fix multiple-monitor strut method for openbox
patching file src/PanelWindow.vala
Hunk #1 succeeded at 379 (offset 210 lines).
Hunk #2 succeeded at 391 (offset 210 lines).
  -> Remove Gala dependecies
patching file src/PanelWindow.vala
Hunk #1 succeeded at 159 with fuzz 2 (offset 73 lines).
Hunk #2 succeeded at 400 with fuzz 2 (offset 249 lines).
Hunk #3 succeeded at 404 with fuzz 1 (offset 243 lines).
patching file src/Widgets/Panel.vala
Hunk #1 succeeded at 25 with fuzz 2.
Hunk #2 succeeded at 79 with fuzz 1 (offset 16 lines).
Hunk #3 succeeded at 265 (offset 37 lines).
patching file src/meson.build
Hunk #1 FAILED at 1.
1 out of 1 hunk FAILED -- saving rejects to file src/meson.build.rej
==> ERROR: A failure occurred in prepare().
    Aborting...
error making: wingpanel-standalone-git</pre>

DAC324 commented:
> _short version_ *****:
> 
> ```
> git clone https://github.com/quequotion/wingpanel.git
> cd wingpanel
> meson build
> ninja build
> ninja -C build install
> ```
> 
As already mentioned by others, the ```ninja build``` step fails with error
<pre>
ninja: error: loading 'build.ninja': No such file or directory</pre>
However, it is possible to install using ```ninja -C build install```. 
Interesting aspect: After the install, wingpanel requires ```libwingpanel-3.0.so.0``` instead of ```libwingpanel-2.0.so.0```. But as I could not find any package providing ```libwingpanel-3.0.so.0```, I just linked my existing version of this library:<pre>sudo ln -sf /usr/lib/libwingpanel-2.0.so.0 /usr/lib/libwingpanel-3.0.so.0</pre> with no problems so far.

quequotion commented:
@DAC324 
>The Wingpanel-standalone-git package appears to be not maintained

It is maintained, I am the maintainer.

Last update is just three months ago.

VCS packages (-git in this case) are only updated when there's a problem with the packaging, not the version of the software available from the source (pkgver bumps happen automatically when the package is built by a user).

I haven't had that problem, but I will look into it.

DAC324 commented:
> At the moment, my AUR package is going to give you a warning because I left `libwingpanel-2.0` in provides, but makepkg is smart and can see it isn't there. What this means is that any indicators (like the stable releases povided in `[community]`) based on libwingpanel-2.0 will not be usable.

The problem with removing `libwingpanel-2.0` is that there seems to be no package other than `wingpanel` providing this file. 
But in order to have Pantheon installed using the metapackages, `libwingpanel-2.0` is required, otherwise, installation of the entire Pantheon package fails prematurely.
Hence, you will have to install `wingpanel` under all circumstances. 

Same with `libwingpanel-3.0`. I was not able to find any package that provides `libwingpanel-3.0`.

quequotion commented:
>using the metapackages

If you are referring to any of [these packages](https://aur.archlinux.org/packages/?O=0&SeB=nd&K=pantheon+meta&outdated=&SB=n&SO=a&PP=50&do_Search=Go) you should know that **none** of them are recommended. I would be _astonished_ to find that any of them work. Do not use any of those packages, please. Just ignore them.

I just took over `pantheon-session-git` to make some [significant changes](https://aur.archlinux.org/packages/pantheon-session-git/#pinned-767181), and updated the Archwiki accordingly. I could use some feedback on that--but preferably either on that wiki's [talk page](https://wiki.archlinux.org/index.php/Talk:Pantheon) or in the Archlinux [forum thread](https://bbs.archlinux.org/viewtopic.php?id=152930) and not here.

tavvfiq commented:
hello, is there any animation on hiding and showing panel?. because i got no animation at all..

quequotion commented:
@tavvfiq 

It should slide up and down, but it's pretty fast (100ms).

tavvfiq commented:
> @tavvfiq
> 
> It should slide up and down, but it's pretty fast (100ms).

yes i see on your demo video. but i cant get the animation to work (just flickering when show/hide)

ylluminate commented:
@quequotion does your fork add notification support? Unfortunately I hit a scenario where I have to have TeamViewer installed on several elementary OS systems and being the company they are, they won't modify the notification / "system tray" methodology and feel that it's a vital part of their service/software. It appears I have to somehow bring the old global notification component back into Hera since it is confusing users with the messages.

quequotion commented:
@ylluminate

No, this pull request has nothing to do with indicators. 

You're looking for wingpanel-indicator-ayatana (and community fork thereof), which was discontinued.

ylluminate commented:
Oh wow, thanks for letting me know @quequotion. That is truly terrible news, but thank you for the name pointer and it does seem like there is still an option for those of us for whom this functionality is critically important: https://github.com/Lafydev/wingpanel-indicator-ayatana

quequotion commented:
At long last, I've had a chance to try this with Gala.

All along, I've been using this with Compiz and Openbox, where it works quite well.

Somehow, the panel never drops down--even with autohide disabled--although the *struts* do.

If autohide is disabled, it seems to bounce meekly in the first two or three rows of pixels at the top of the screen on mouseover, but never comes down. Meanwhile, the struts *do* push windows out of the panel space for autohide or dodge modes.

I can't see why this is happening, especially since it works with other window managers, but the most suspicious code to me is what I borrowed from @donadigo's implementation, offloading the animation to the panel widget. It's possible I left out something needed especially by Gala or that Gala has changed in some way since it was written; but to be honest I have no idea if it really has anything to do with it.

n-osennij commented:
4 years, no result((

id-ekaagr commented:
Whats's the status?
